internal/pkg/cli/deploy.go

```diff
@@ -208,6 +276,146 @@ func (o *deployOpts) askName() error {
+	o.envExistsInApp = true
+	_, err := o.store.GetEnvironment(o.appName, o.envName)
+	if err != nil {
+		var errNotFound *config.ErrNoSuchEnvironment
+		if !errors.As(err, &errNotFound) {
+			return fmt.Errorf("get environment from config store: %w", err)
+		}
+		o.envExistsInApp = false
+	}
+	envs, err := o.ws.ListEnvironments()
+	if err != nil {
+		return fmt.Errorf("list environments in workspace: %w", err)
+	}
+	o.envExistsInWs = contains(o.envName, envs)
+
+	// the desired environment doesn't actually exist.
+	if !o.envExistsInApp && !o.envExistsInWs {
+		log.Errorf("Environment %q does not exist in the current application or workspace. Please initialize it by running %s.\n", o.envName, color.HighlightCode("copilot env init"))
+		return fmt.Errorf("environment %q does not exist in the workspace", o.envName)
+	}
+	if o.envExistsInApp && !o.envExistsInWs {
+		log.Infof("Manifest for environment %q does not exist in the current workspace. To deploy this environment, generate a manifest with %s", o.envName, color.HighlightCode("copilot env show --manifest"))
+	}
+
+	return nil
+}
+
+func (o *deployOpts) maybeInitEnv() error {
+	// Env has no manifest and doesn't exist in app; we can't do anything more.
+	if !o.envExistsInWs && !o.envExistsInApp {
+		return fmt.Errorf("environment %q does not exist in the workspace", o.envName)
+	}
+
+	if o.envExistsInApp {
+		log.Infof("Environment %q exists in application %q. Skipping initialization.\n", o.envName, o.appName)
+		return nil
+	}
+
+	// If no initialization flags were specified and the env wasn't initialized, ask to confirm.
+	if !o.envExistsInApp && o.yesInitEnv == nil {
+		v, err := o.prompt.Confirm(fmt.Sprintf("Environment %q does not exist in app %q. Initialize it?", o.envName, o.appName), "")
+		if err != nil {
+			return fmt.Errorf("confirm env init: %w", err)
+		}
+		o.yesInitEnv = aws.Bool(v)
+	}
+
+	if aws.BoolValue(o.yesInitEnv) {
+		cmd, err := o.newInitEnvCmd(o)
+		if err != nil {
+			return fmt.Errorf("load env init command : %w", err)
+		}
+		if err = cmd.Validate(); err != nil {
+			return err
+		}
+		if err = cmd.Ask(); err != nil {
+			return err
+		}
+		if err = cmd.Execute(); err != nil {
+			return err
+		}
+		if o.deployEnv == nil {
+			log.Infof("Environment %q was just initialized. We'll deploy it now.\n", o.envName)
+			o.deployEnv = aws.Bool(true)
+		} else if !aws.BoolValue(o.deployEnv) {
+			log.Errorf("Environment is not deployed but --%s=false was specified. Deploy the environment with %s in order to deploy a workload to it.\n", deployEnvFlag, color.HighlightCode("copilot env deploy"))
+			return fmt.Errorf("environment %s was initialized but has not been deployed", o.envName)
+		}
+		return nil
+	}
+	log.Errorf("Environment %q does not exist in application %q and was not initialized after prompting.\n", o.envName, o.appName)
+	return fmt.Errorf("env %s does not exist in app %s", o.envName, o.appName)
+}
+
+func (o *deployOpts) maybeDeployEnv() error {
```

Comment:
I think one case that this does not handle is this:
```
$copilot deploy
Which env?
> test (initted, but not in workspace and not-yet deployed)

# Error when trying to deploy svc to "test", because the test env isn't deployed yet.
```

I think it is fine to not handle this case, at least not for this release though! I just want to call out that this _might_ be something that we want to support. But again, not a blocking issue at all.

Reply:
understood. I think there may be a big refactor of the init and deploy experiences coming soonish and we could better support this case. as of right now though I don't think we have special checks in place to see whether the env is initted but not deployed even in `copilot svc deploy` or the like.

Reply:
It is done here! https://github.com/aws/copilot-cli/blob/7ae7953cb891287c6de83ba3daa7354a3db36bfb/internal/pkg/cli/svc_deploy.go#L511

There were a few changes regarding what `env init` do in the last year. `env init`, in addition to create SSM parameters, it also deploys an environment stack that contains _only) env manager role and cfn execution role - sort of like the CDK bootstrap stack.

The bootstrap stack's template "version" is a special string and can be used to determine the state of an environment.

This is just a note for future refactor! Not a request to update this PR :shipit: